File: androidApp/src/main/kotlin/it/mensa/app/features/sigs/AddSigSheet.kt
package it.mensa.app.features.sigs

import android.net.Uri
import androidx.activity.compose.rememberLauncherForActivityResult
import androidx.activity.result.contract.ActivityResultContracts
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.outlined.AddPhotoAlternate
import androidx.compose.material.icons.outlined.Close
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.SolidColor
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import it.mensa.app.support.FilesUrl
import it.mensa.app.support.tr
import it.mensa.app.ui.components.CachedAsyncImage
import it.mensa.shared.model.SigModel
import it.mensa.shared.repository.SigDraft

enum class SigGroupType(val rawValue: String, val label: String) {
    SigFacebook("sig_facebook", "SIG Facebook"),
    SigGeneric("sig", "SIG Generic"),
    Local("local", "Gruppo locale"),
    ChatWhatsapp("chat_whatsapp", "Chat WhatsApp"),
    ChatTelegram("chat_telegram", "Chat Telegram"),
    Chat("chat", "Chat");

    companion object {
        fun fromRaw(raw: String): SigGroupType =
            values().firstOrNull { it.rawValue == raw } ?: SigGeneric
    }
}

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun AddSigSheet(
    initial: SigModel?,
    onSubmitted: (SigDraft) -> Unit,
    onDeleteRequested: () -> Unit = {},
    onDismiss: () -> Unit,
) {
    val isEditing = initial != null
    val context = LocalContext.current

    var name by remember { mutableStateOf(initial?.name ?: "") }
    var link by remember { mutableStateOf(initial?.link ?: "") }
    var groupType by remember {
        mutableStateOf(SigGroupType.fromRaw(initial?.groupType ?: "sig"))
    }
    var imageUri by remember { mutableStateOf<Uri?>(null) }
    var imageBytes by remember { mutableStateOf<ByteArray?>(null) }
    var showDeleteConfirm by remember { mutableStateOf(false) }

    val canSubmit = name.trim().isNotEmpty() && link.trim().isNotEmpty()

    val imagePickerLauncher = rememberLauncherForActivityResult(
        contract = ActivityResultContracts.GetContent()
    ) { uri ->
        if (uri != null) {
            imageUri = uri
            imageBytes = context.contentResolver.openInputStream(uri)?.use { it.readBytes() }
        }
    }

    val remoteCoverUrl = remember(initial) {
        if (initial == null || initial.image.isEmpty()) null
        else if (initial.image.startsWith("http")) initial.image
        else FilesUrl.build("sigs", initial.id, initial.image, "1500x600")
    }

    ModalBottomSheet(
        onDismissRequest = onDismiss,
        sheetState = rememberModalBottomSheetState(skipPartiallyExpanded = true),
    ) {
        Column(
            modifier = Modifier
                .fillMaxWidth()
                .verticalScroll(rememberScrollState())
                .padding(horizontal = 16.dp)
                .padding(bottom = 32.dp),
            verticalArrangement = Arrangement.spacedBy(16.dp),
        ) {
            // Header
            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.SpaceBetween,
                verticalAlignment = Alignment.CenterVertically,
            ) {
                TextButton(onClick = onDismiss) {
                    Text(tr("app.cancel", fallback = "Annulla"))
                }
                Text(
                    if (isEditing)
                        tr("sigs.edit.title", fallback = "Modifica community")
                    else
                        tr("sigs.add.title", fallback = "Nuova community"),
                    style = MaterialTheme.typography.titleMedium,
                    fontWeight = FontWeight.SemiBold,
                )
                TextButton(
                    onClick = {
                        if (!canSubmit) return@TextButton
                        val draft = SigDraft(
                            name = name.trim(),
                            link = link.trim(),
                            groupType = groupType.rawValue,
                            description = "",
                            imageBytes = imageBytes,
                            imageFilename = if (imageBytes != null) "cover.jpg" else null,
                            imageContentType = if (imageBytes != null) "image/jpeg" else null,
                        )
                        onSubmitted(draft)
                    },
                    enabled = canSubmit,
                ) {
                    Text(
                        if (isEditing)
                            tr("app.update", fallback = "Aggiorna")
                        else
                            tr("app.create", fallback = "Crea"),
                        fontWeight = FontWeight.SemiBold,
                    )
                }
            }

            // Cover picker
            Box(
                modifier = Modifier
                    .fillMaxWidth()
                    .aspectRatio(1528f / 603f)
                    .clip(RoundedCornerShape(22.dp))
                    .clickable { imagePickerLauncher.launch("image/*") },
            ) {
                when {
                    imageUri != null -> {
                        CachedAsyncImage(
                            model = imageUri,
                            contentDescription = "Copertina",
                            modifier = Modifier.fillMaxSize(),
                            contentScale = ContentScale.Crop,
                        )
                        IconButton(
                            onClick = { imageUri = null; imageBytes = null },
                            modifier = Modifier.align(Alignment.TopEnd).padding(8.dp),
                        ) {
                            Icon(
                                Icons.Outlined.Close,
                                contentDescription = "Rimuovi",
                                tint = Color.White,
                            )
                        }
                    }
                    remoteCoverUrl != null -> {
                        CachedAsyncImage(
                            model = remoteCoverUrl,
                            contentDescription = "Copertina",
                            modifier = Modifier.fillMaxSize(),
                            contentScale = ContentScale.Crop,
                        )
                    }
                    else -> {
                        Box(
                            modifier = Modifier
                                .fillMaxSize()
                                .background(MaterialTheme.colorScheme.surfaceVariant)
                                .border(
                                    width = 1.5.dp,
                                    brush = SolidColor(MaterialTheme.colorScheme.outline.copy(alpha = 0.5f)),
                                    shape = RoundedCornerShape(22.dp),
                                ),
                            contentAlignment = Alignment.Center,
                        ) {
                            Column(
                                horizontalAlignment = Alignment.CenterHorizontally,
                                verticalArrangement = Arrangement.spacedBy(6.dp),
                            ) {
                                Icon(
                                    Icons.Outlined.AddPhotoAlternate,
                                    contentDescription = null,
                                    tint = MaterialTheme.colorScheme.onSurfaceVariant,
                                )
                                Text(
                                    tr("sigs.cover.pick", fallback = "Tocca per scegliere una copertina"),
                                    style = MaterialTheme.typography.bodySmall,
                                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                                )
                            }
                        }
                    }
                }
            }

            // Name field
            OutlinedTextField(
                value = name,
                onValueChange = { name = it },
                label = { Text(tr("sigs.field.name", fallback = "Nome")) },
                modifier = Modifier.fillMaxWidth(),
                singleLine = true,
            )

            // Link field
            OutlinedTextField(
                value = link,
                onValueChange = { link = it },
                label = { Text(tr("sigs.field.link", fallback = "Link")) },
                modifier = Modifier.fillMaxWidth(),
                singleLine = true,
            )

            // Type picker
            Text(
                tr("sigs.section.type", fallback = "Tipo"),
                style = MaterialTheme.typography.labelLarge,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
            )
            SigGroupType.values().forEach { type ->
                Row(
                    modifier = Modifier
                        .fillMaxWidth()
                        .clickable { groupType = type }
                        .padding(vertical = 4.dp),
                    verticalAlignment = Alignment.CenterVertically,
                    horizontalArrangement = Arrangement.spacedBy(8.dp),
                ) {
                    RadioButton(
                        selected = groupType == type,
                        onClick = { groupType = type },
                    )
                    Text(type.label, style = MaterialTheme.typography.bodyMedium)
                }
            }

            // Delete button (edit only)
            if (isEditing) {
                Divider()
                TextButton(
                    onClick = { showDeleteConfirm = true },
                    modifier = Modifier.fillMaxWidth(),
                    colors = ButtonDefaults.textButtonColors(contentColor = MaterialTheme.colorScheme.error),
                ) {
                    Text(tr("sigs.delete", fallback = "Elimina community"))
                }
            }
        }
    }

    if (showDeleteConfirm) {
        AlertDialog(
            onDismissRequest = { showDeleteConfirm = false },
            title = { Text(tr("sigs.delete.confirm.title", fallback = "Eliminare questa community?")) },
            text = { Text(tr("sigs.delete.confirm.body", fallback = "Questa azione non è annullabile.")) },
            confirmButton = {
                TextButton(
                    onClick = {
                        showDeleteConfirm = false
                        onDeleteRequested()
                    },
                    colors = ButtonDefaults.textButtonColors(contentColor = MaterialTheme.colorScheme.error),
                ) {
                    Text(tr("sigs.delete", fallback = "Elimina community"))
                }
            },
            dismissButton = {
                TextButton(onClick = { showDeleteConfirm = false }) {
                    Text(tr("app.cancel", fallback = "Annulla"))
                }
            },
        )
    }
}

File: iosApp/iosApp/Features/Sigs/AddSigSheet.swift
import SwiftUI
import PhotosUI
import Shared

/// Group-type keys recognised by the Mensa Italia backend.
/// Order in the picker matches the Flutter app verbatim.
enum SigGroupType: String, CaseIterable, Identifiable {
    case sigFacebook  = "sig_facebook"
    case sigGeneric   = "sig"
    case local        = "local"
    case chatWhatsapp = "chat_whatsapp"
    case chatTelegram = "chat_telegram"
    case chat         = "chat"

    var id: String { rawValue }

    /// Human-readable label (Italian fallback) — uses `tr(key, fallback:)`.
    var label: String {
        switch self {
        case .sigFacebook:  return tr("sigs.type.sig_facebook",  fallback: "SIG Facebook")
        case .sigGeneric:   return tr("sigs.type.sig",           fallback: "SIG Generic")
        case .local:        return tr("sigs.type.local",         fallback: "Gruppo locale")
        case .chatWhatsapp: return tr("sigs.type.chat_whatsapp", fallback: "Chat WhatsApp")
        case .chatTelegram: return tr("sigs.type.chat_telegram", fallback: "Chat Telegram")
        case .chat:         return tr("sigs.type.chat",          fallback: "Chat")
        }
    }

    /// SF Symbol for the type — used as a leading glyph in the picker and list rows elsewhere.
    var systemImage: String {
        switch self {
        case .sigFacebook:  return "f.circle.fill"
        case .sigGeneric:   return "person.3.fill"
        case .local:        return "mappin.and.ellipse"
        case .chatWhatsapp: return "bubble.left.and.bubble.right.fill"
        case .chatTelegram: return "paperplane.fill"
        case .chat:         return "message.fill"
        }
    }

    init?(rawString: String) { self.init(rawValue: rawString) }
}

/// Payload returned by `AddSigSheet` to its caller.
/// `imageData == nil` in edit mode means "keep existing image".
struct SigDraftPayload {
    var name: String
    var link: String
    var groupType: SigGroupType
    var imageData: Data?
    var imageFilename: String?
    var imageContentType: String?
}

struct AddSigSheet: View {
    /// nil → create. non-nil → edit (form is pre-populated).
    var initial: SigModel?

    /// Called when the user taps "Crea" / "Aggiorna" with the assembled payload.
    /// Caller is responsible for persisting via `koin.sigs.create/update` and dismissing.
    var onSubmitted: (SigDraftPayload) -> Void

    /// Called when the user taps the destructive delete CTA (only present in edit mode).
    /// Caller persists the delete via `koin.sigs.delete(id:)` and dismisses.
    var onDeleteRequested: () -> Void = {}

    /// Called on cancel. Default no-op.
    var onCancelled: () -> Void = {}

    @State private var name: String
    @State private var link: String
    @State private var groupType: SigGroupType
    @State private var imageData: Data?
    @State private var imageFilename: String?
    @State private var imageContentType: String?
    @State private var pickerItem: PhotosPickerItem?
    @State private var showPhotoPicker: Bool = false
    @State private var showDeleteConfirm: Bool = false

    init(
        initial: SigModel? = nil,
        onSubmitted: @escaping (SigDraftPayload) -> Void,
        onDeleteRequested: @escaping () -> Void = {},
        onCancelled: @escaping () -> Void = {}
    ) {
        self.initial = initial
        self.onSubmitted = onSubmitted
        self.onDeleteRequested = onDeleteRequested
        self.onCancelled = onCancelled

        _name = State(initialValue: initial?.name ?? "")
        _link = State(initialValue: initial?.link ?? "")
        let seedType: SigGroupType = {
            if let raw = initial?.groupType, let t = SigGroupType(rawValue: raw) {
                return t
            }
            return .sigGeneric
        }()
        _groupType = State(initialValue: seedType)
        _imageData = State(initialValue: nil)
        _imageFilename = State(initialValue: nil)
        _imageContentType = State(initialValue: nil)
    }

    private var isEditing: Bool { initial != nil }

    private var trimmedName: String {
        name.trimmingCharacters(in: .whitespacesAndNewlines)
    }
    private var trimmedLink: String {
        link.trimmingCharacters(in: .whitespacesAndNewlines)
    }
    private var canSubmit: Bool {
        !trimmedName.isEmpty && !trimmedLink.isEmpty
    }

    var body: some View {
        NavigationStack {
            Form {
                Section {
                    coverPicker
                        .listRowInsets(EdgeInsets(top: 8, leading: 8, bottom: 8, trailing: 8))
                        .listRowBackground(Color.clear)
                }

                Section(tr("sigs.section.details", fallback: "Dettagli")) {
                    TextField(
                        tr("sigs.field.name", fallback: "Nome"),
                        text: $name
                    )
                    TextField(
                        tr("sigs.field.link", fallback: "Link"),
                        text: $link
                    )
                    .keyboardType(.URL)
                    .textInputAutocapitalization(.never)
                    .autocorrectionDisabled()
                }

                Section(tr("sigs.section.type", fallback: "Tipo")) {
                    Picker(
                        tr("sigs.section.type", fallback: "Tipo"),
                        selection: $groupType
                    ) {
                        ForEach(SigGroupType.allCases) { type in
                            Label(type.label, systemImage: type.systemImage)
                                .tag(type)
                        }
                    }
                    .pickerStyle(.menu)
                }

                if isEditing {
                    Section {
                        Button(role: .destructive) {
                            showDeleteConfirm = true
                        } label: {
                            Text(tr("sigs.delete", fallback: "Elimina community"))
                                .frame(maxWidth: .infinity, alignment: .center)
                        }
                    }
                }
            }
            .navigationTitle(
                isEditing
                    ? tr("sigs.edit.title", fallback: "Modifica community")
                    : tr("sigs.add.title", fallback: "Nuova community")
            )
            .navigationBarTitleDisplayMode(.inline)
            .toolbar {
                ToolbarItem(placement: .topBarLeading) {
                    Button(tr("app.cancel", fallback: "Annulla")) {
                        onCancelled()
                    }
                }
                ToolbarItem(placement: .topBarTrailing) {
                    Button(
                        isEditing
                            ? tr("app.update", fallback: "Aggiorna")
                            : tr("app.create", fallback: "Crea")
                    ) {
                        submit()
                    }
                    .disabled(!canSubmit)
                }
            }
            .photosPicker(
                isPresented: $showPhotoPicker,
                selection: $pickerItem,
                matching: .images,
                photoLibrary: .shared()
            )
            .onChange(of: pickerItem) { _, newItem in
                guard let item = newItem else { return }
                Task {
                    if let data = try? await item.loadTransferable(type: Data.self) {
                        await MainActor.run {
                            imageData = data
                            imageFilename = "cover.jpg"
                            imageContentType = "image/jpeg"
                        }
                    }
                    await MainActor.run { pickerItem = nil }
                }
            }
            .confirmationDialog(
                tr("sigs.delete.confirm.title", fallback: "Eliminare questa community?"),
                isPresented: $showDeleteConfirm,
                titleVisibility: .visible
            ) {
                Button(
                    tr("sigs.delete", fallback: "Elimina community"),
                    role: .destructive
                ) {
                    onDeleteRequested()
                }
                Button(tr("app.cancel", fallback: "Annulla"), role: .cancel) { }
            } message: {
                Text(tr(
                    "sigs.delete.confirm.body",
                    fallback: "Questa azione non è annullabile."
                ))
            }
        }
    }

    // MARK: - Cover picker

    @ViewBuilder
    private var coverPicker: some View {
        Button {
            showPhotoPicker = true
        } label: {
            coverContent
        }
        .buttonStyle(.plain)
    }

    @ViewBuilder
    private var coverContent: some View {
        ZStack {
            if let data = imageData, let ui = UIImage(data: data) {
                Image(uiImage: ui)
                    .resizable()
                    .scaledToFill()
            } else if let url = remoteCoverURL() {
                CachedAsyncImage(url: url) { img in
                    img.resizable().scaledToFill()
                } placeholder: {
                    placeholderBackground
                }
            } else {
                placeholderBackground
                VStack(spacing: 6) {
                    Image(systemName: "photo.badge.plus")
                        .font(.title2)
                        .foregroundStyle(.secondary)
                    Text(tr("sigs.cover.pick", fallback: "Tocca per scegliere una copertina"))
                        .font(.subheadline)
                        .foregroundStyle(.secondary)
                        .multilineTextAlignment(.center)
                        .padding(.horizontal, 16)
                }
            }

            if isEditing, imageData != nil {
                VStack {
                    HStack {
                        Spacer()
                        Button {
                            imageData = nil
                            imageFilename = nil
                            imageContentType = nil
                        } label: {
                            Label(
                                tr("sigs.cover.remove", fallback: "Rimuovi"),
                                systemImage: "xmark.circle.fill"
                            )
                            .labelStyle(.titleAndIcon)
                            .font(.caption.weight(.semibold))
                            .padding(.horizontal, 10)
                            .padding(.vertical, 6)
                            .background(.ultraThinMaterial, in: Capsule())
                            .foregroundStyle(.primary)
                        }
                        .buttonStyle(.plain)
                        .padding(10)
                    }
                    Spacer()
                }
            }
        }
        .aspectRatio(1528.0 / 603.0, contentMode: .fit)
        .frame(maxWidth: .infinity)
        .clipShape(RoundedRectangle(cornerRadius: 22, style: .continuous))
        .contentShape(RoundedRectangle(cornerRadius: 22, style: .continuous))
    }

    @ViewBuilder
    private var placeholderBackground: some View {
        ZStack {
            RoundedRectangle(cornerRadius: 22, style: .continuous)
                .fill(Color(.secondarySystemBackground))
            RoundedRectangle(cornerRadius: 22, style: .continuous)
                .strokeBorder(
                    Color.secondary.opacity(0.5),
                    style: StrokeStyle(lineWidth: 1.5, dash: [4, 4])
                )
        }
    }

    private func remoteCoverURL() -> URL? {
        guard let sig = initial, !sig.image.isEmpty else { return nil }
        if sig.image.hasPrefix("http") { return URL(string: sig.image) }
        return Files.url(
            collection: "sigs",
            recordId: sig.id,
            filename: sig.image,
            thumb: "1500x600"
        )
    }

    // MARK: - Submit

    private func submit() {
        guard canSubmit else { return }
        let payload = SigDraftPayload(
            name: trimmedName,
            link: trimmedLink,
            groupType: groupType,
            imageData: imageData,
            imageFilename: imageFilename,
            imageContentType: imageContentType
        )
        onSubmitted(payload)
    }
}
